### PR TITLE
flamenco, runtime: support multiple epoch boundaries

### PR DESCRIPTION
**Pull request by topointon-jump**
This PR fixes several subtle bugs in how the epoch stakes cache is propagated during epoch boundaries. It's easiest to review commit-by-commit.

Also adds a series of 5-node ledger tests, which go through 5 consecutive epochs in a single execution and have hundreds of thousands of stake accounts.

### CODE REVIEW

File: src/flamenco/runtime/context/fd_exec_slot_ctx.c [hunk 1]
```diff
@@ -329,11 +329,24 @@ fd_exec_slot_ctx_recover_( fd_exec_slot_ctx_t *   slot_ctx,
     fd_epoch_epoch_stakes_pair_t * epochs  = oldbank->epoch_stakes;
     fd_epoch_stakes_t *            stakes0 = NULL;  /* current */
     fd_epoch_stakes_t *            stakes1 = NULL;  /* next */
-    for( ulong i=0UL; i < manifest->bank.epoch_stakes_len; i++ ) {
+    slot_ctx->slot_bank.has_use_preceeding_epoch_stakes = 1;
+    slot_ctx->slot_bank.use_preceeding_epoch_stakes = epoch + 2UL;
```
Comment (ibhatt-jumptrading) [inline src/flamenco/runtime/context/fd_exec_slot_ctx.c:333]: nit: align equals

File: src/flamenco/runtime/context/fd_exec_slot_ctx.c [hunk 2]
```diff
@@ -329,11 +329,24 @@ fd_exec_slot_ctx_recover_( fd_exec_slot_ctx_t *   slot_ctx,
...
         stakes1 = &epochs[i].value;
+
+      /* When loading from a snapshot, Agave's stake caches mean that we have to special-case the epoch stakes
+         that are used for the second epoch E+2 after the snapshot epoch E.
+         
+         If the snapshot contains the epoch stakes for E+2, we should use those.
+         
+         If the snapshot does not, we should use the stakes at the end of the E-1 epoch, instead of E-2 as we do for
+         all other epochs. */
+      if ( epochs[i].key == epoch+2UL )
```
Comment (ibhatt-jumptrading) [inline src/flamenco/runtime/context/fd_exec_slot_ctx.c:348]: nit parents + brackets

File: src/flamenco/runtime/fd_runtime.c [hunk 3]
```diff
@@ -3710,6 +3708,20 @@ void fd_update_next_epoch_stakes( fd_exec_slot_ctx_t * slot_ctx ) {
...
+
+  In this function:
+  - leader schedule should be calculated using T-2 stakes (slot_ctx->slot_bank.epoch_stakes)
+  - stakes in T-2 (slot_ctx->slot_bank.epoch_stakes) should be replaced by T-1 (epoch_bank->next_epoch_stakes)
+  - stakes at T-1 (epoch_bank->next_epoch_stakes) should be replaced by updated stakes at T (stakes->vote_accounts)
+
+  Invariant during an epoch T:
+  epoch_bank->next_epoch_stakes    holds the stakes at T-1
+  slot_ctx->slot_bank.epoch_stakes holds the stakes at T-2
+ */
```
Comment (jumpsiegel) [inline src/flamenco/runtime/fd_runtime.c:3724]: So, the manifest only needs T-2 of stakes... even through they stick in 5 of them?

**Review by jumpsiegel — APPROVED** [review]

File: src/flamenco/runtime/fd_runtime.c (hunk 3 above)
Comment (ibhatt-jumptrading) [reply src/flamenco/runtime/fd_runtime.c:3724]: would be nice to leave a comment in the code explicitly explaining this

Comment (topointon-jump) [reply src/flamenco/runtime/fd_runtime.c:3724]: @jumpsiegel yes exactly, I'm not sure exactly why they store 5 as you can always derive them if you have 2 preceeding epochs

File: src/flamenco/runtime/sysvar/fd_sysvar_cache.c [hunk 4]
```diff
@@ -86,43 +86,52 @@ FD_SYSVAR_CACHE_ITER(X)
 
 /* Restore sysvars */
 
+# define X( type, name )                                                  \
```
Comment (ibhatt-jumptrading) [inline src/flamenco/runtime/sysvar/fd_sysvar_cache.c:89]: These macros can be kind of unclear to read through, but don't feel super strongly about these. Also some style nits here

**Review by ibhatt-jumptrading — CHANGES_REQUESTED** [review]
mostly lgtm, just some style concerns

File: src/flamenco/runtime/sysvar/fd_sysvar_cache.c (hunk 4 above)
Comment (topointon-jump) [reply src/flamenco/runtime/sysvar/fd_sysvar_cache.c:89]: I just copied the existing style here in this file

**Review by ibhatt-jumptrading — DISMISSED** [review]

**Review by ibhatt-jumptrading — APPROVED** [review]

**Pull request merged**